content/blog/kube-dns-coredns-perfcomp.md

```diff
@@ -0,0 +1,114 @@
+
+**Kubernetes DNS Perf-test Tool**: I used the [kubernetes dns perf-tests tool](https://github.com/kubernetes/perf-tests/tree/master/dns) modified to [allow tests on CoreDNS](https://github.com/kubernetes/perf-tests/pull/114).  The tool by default runs the DNS server under test on the master node and the client pod on separate node (node 1).  The tests are not performed against the cluster's active cluster DNS, instead they configure, spin up and tear down an instance for each set of tests.
+
+**Upstream Server**: The upstream server was a CoreDNS instance using the `template` plugin, configured to reply to all queries with answer of '1.2.3.4'.  It was run in the cluster, on a separate node from the client and DNS server under test (node 2).  This local server responds much more quickly than a real world upstream server would.
+
+**DNS Server Versions**: CoreDNS 1.1.3, kube-dns 1.14.10
+
+
+## CoreDNS and Kube-dns QPS and Latency
+
+Each set of tests is for 3 types of `A` record queries:
+
+1. **invalid**: A name for a service that does not exist in the cluster, which prompts an `NXDOMAIN` response from the DNS service. e.g. `invalid-service.default.svc.cluster.local`
+2. **service**: A name for a service that exists in the cluster. e.g. `kubernetes.default.svc.cluster.local`
+3. **upstream**: A name outside of the cluster, which is forwarded upstream by the DNS service. e.g. `coredns.io`
+
+
+### Comparisons on a Level(ish) Playing Field
+
+Dnsmasq is single threaded, so when running on a multi core system, it will use at most one core. CoreDNS is multi threaded, so it could use all cores.  This presents an unfair advantage to CoreDNS when comparing single instances on a multi core system.  Scaling kube-dns instances horizontally in the test to level the playing field would be difficult to tune evenly.  So instead, I ran tests with restricted CPU allotted per container, such that the equivalent of one core at maximum would be used by the dns servers during the tests. 
+
+Since CoreDNS operates in a single container, restricting CPU is easy.  But kube-dns contains two containers, so ensuring both its containers operate within one core without starving one is not as straight forward.  To account for this, I ran kube-dns CPU restrictions in two separate runs, one with all CPU allocated to dnsmasq (for tests of dnsmasq only tasks e.g. cache hits and upstream forwarding), and a second with kube-dns and dnsmasq sharing the CPU at it's nominal ratio (observationally, 70% kube-dns, 30% dnsmasq).
+
+| Query     | kube-dns QPS | CoreDNS QPS | kube-dns LAT | CoreDNS LAT |
+| :---      |     ---:     |      ---:   |     ---:     |      ---:   |
+| invalid   |     2775     |      5019   |     36.0     |      19.9   |
```

Comment:
These latencies seem really high based on my experience.

Reply:
OK.  How would you suggest I proceed?

Reply:
(i.e. do you think there is something I need to do/change here?)

Reply:
Well, I am wondering if this is purely a CPU issue. We know we can do more QPS than that in many circumstances. And our latencies in tests running not in Docker were 1ms or less.

Reply:
I did limit the execution to one core.

Reply:
What, other than CPU could it be?

Reply:
I dont think it's a network thing, since the latency is so low for dnsmasq...

Reply:
The goal of the blog was to compare kube-dns to CoreDNS on a level playing field to see what the relative performance impact would be when migrating from kube-dns to CoreDNS.

The goal was not to showcase CoreDNS numbers on their own... however, in retrospect I think most people will probably not actually read this blog, and just skip to the results, in which case they may think "wow coredns has bad numbers"... being ignorant of the context.